Add direct calls to the FlowEVM doc (#760)
onflow/docs: docs/evm/cadence/direct-calls.md

---

> title: Direct Calls from Cadence to FlowEVM

Reviewer: ```suggestion
title: Direct Calls from Cadence to Flow EVM
```

> title: Direct Calls from Cadence to the FlowEVM

Reviewer: ```suggestion
title: Direct Calls from Cadence to Flow EVM
```

> sidebar_label: Direct Calls to FlowEVM

Reviewer: ```suggestion
sidebar_label: Direct Calls to Flow EVM
```

> sidebar_label: Direct Calls to FlowEVM

Reviewer: ```suggestion
sidebar_label: Direct Calls to Flow EVM
```

sidebar_position: 5
---

> Direct calls from Cadence to the FlowEVM are essential for enabling Cadence smart contracts to interact seamlessly with the EVM environment hosted on the Flow blockchain. These calls facilitate a range of functionalities including state queries and transaction initiations, allowing Cadence contracts to leverage EVM-based tools and assets.

Reviewer: ```suggestion
Direct calls from Cadence to Flow EVM are essential for enabling Cadence smart contracts to interact seamlessly with the EVM environment hosted on the Flow blockchain. These calls facilitate a range of functionalities including state queries and transaction initiations, allowing Cadence contracts to leverage EVM-based tools and assets.
```

## Making Direct Calls

> ### Accessing the FlowEVM

Reviewer: ```suggestion
### Accessing Flow EVM
```

> To interact with the FlowEVM, Cadence contracts must first import `EVM` from its service address:

Reviewer: ```suggestion
To interact with Flow EVM, Cadence contracts must first import `EVM` from its service address:
```

```js
import EVM from <ServiceAddress>
```

Next, create an `EVMAddress` with a sequence of 20 bytes representing the EVM address:

```js
let addr = EVM.EVMAddress(bytes: bytes)
```

Once you have access to an `EVMAddress`, you can query various pieces of state information such as:

- `balance() EVM.Balance` provides the balance of the address. It returns a balance object rather than a basic type to avoid errors when converting from flow to atto-flow.
- `nonce() UInt64`  retrieves the nonce associated with the address.
- `code(): [UInt8]` fetches the code at the address; it returns the smart contract code if applicable, and is empty otherwise.

```
import EVM from <ServiceAddress>

access(all)
fun main(bytes: [UInt8; 20]) {
    let addr = EVM.EVMAddress(bytes: bytes)
    let bal = addr.balance()
}
```

> ### Sending Transactions to the FlowEVM

Reviewer: ```suggestion
### Sending Transactions to Flow EVM
```

> To send transactions to the FlowEVM, use the `run` function which executes RLP-encoded transactions. RLP (Recursive Length Prefix) encoding is used to efficiently encode data into a byte-array format, suitable for Ethereum-based environments. Here's an example of wrapping and sending a transaction:

Reviewer: ```suggestion
To send transactions to Flow EVM, use the `run` function which executes RLP-encoded transactions. RLP (Recursive Length Prefix) encoding is used to efficiently encode data into a byte-array format, suitable for Ethereum-based environments. Here's an example of wrapping and sending a transaction:
```

```
import EVM from <ServiceAddress>

transaction(rlpEncodedTransaction: [UInt8], coinbaseBytes: [UInt8; 20]) {

    prepare(signer: AuthAccount) {
        let coinbase = EVM.EVMAddress(bytes: coinbaseBytes)
        let result = EVM.run(tx: rlpEncodedTransaction, coinbase: coinbase)
        assert(
            runResult.status == Status.successful,
            message: "tx was not executed successfully."
        )
    }
}
```

> Using `run` restricts an EVM block to a single EVM transaction, while a future `batchRun` will offer the capability to execute multiple EVM transactions in a batch.

Reviewer: I believe `batchRun` has been added already

- `Status.invalid`: This status indicates that the transaction or call failed at the validation step, such as due to a nonce mismatch. Transactions with this status are not executed or included in a block, meaning no state change occurs.
- `Status.failed`: This status is assigned when the transaction has technically succeeded in terms of being processable, but the EVM reports an error as the outcome, such as running out of gas. Importantly, a failed transaction or call is still included in a block. Attempting to resubmit a failed transaction will result in an `invalid` status on the second try due to a now incorrect nonce.
- `Status.successful`: This status is given when the transaction or call is successfully executed and no errors are reported by the EVM.

For scenarios where transaction validity is critical, developers may choose to use the `mustRun` variation, which reverts the transaction in the case of a validation failure, providing an added layer of error handling.

### Understanding Gas Usage in EVM Transactions

> Direct calls to the FlowEVM require gas, it's important to understand how gas usage is calculated and billed. During the execution of methods that interact with the EVM:

Reviewer: ```suggestion
Direct calls to Flow EVM require gas, it's important to understand how gas usage is calculated and billed. During the execution of methods that interact with the EVM:
```
